<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import folium
import pandas as pd

In [2]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [3]:
## Task 1: Mark all launch sites on a map
import folium

# Example coordinates for launch sites
launch_sites = {
    'CAFS LC-40': [28.562302,-80.577356],
    'CCAFS SLC-40': [28.563197,-80.576820],
    'KSC LC-39A': [28.573255,-80.646895],
    'VAFB SLC-4E': [34.632834,-120.610745]
}

# Create a map centered roughly at the average location
map_center = [28.5, -80.6]
launch_map = folium.Map(location=map_center, zoom_start=5)

# Add markers for each launch site
for site, coords in launch_sites.items():
    folium.Marker(location=coords, popup=site).add_to(launch_map)

# Display the map
launch_map

First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
import requests
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = requests.get(URL)
spacex_csv_file = io.BytesIO(resp.content)
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Iterate through each launch site and add a Circle and a Marker
for index, row in launch_sites_df.iterrows():
    coordinate = [row['Lat'], row['Long']]
    launch_site_name = row['Launch Site']

    # Add a Circle for the launch site
    folium.Circle(
        coordinate,
        radius=1000,  # Radius in meters
        color='#d35400',
        fill=True,
        fill_color='#d35400',
        fill_opacity=0.6
    ).add_child(folium.Popup(launch_site_name)).add_to(site_map)

    # Add a DivIcon marker for the launch site
    folium.Marker(
        coordinate,
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html=f'<div style="font-size: 12px; color:#d35400;"><b>{launch_site_name}</b></div>'
        )
    ).add_to(site_map)
site_map

In [9]:
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for index, row in launch_sites_df.iterrows():
    folium.Circle(
        location=[row['Lat'],row['Long']],
        radius =1000,
        color='#0000FF',
        fill = True,
        fill_color='#0000FF',
        fill_opacity=0.2
    ).add_child(folium.Popup(row['Launch Site'])).add_to(site_map)
    folium.Marker(
        location=[row['Lat'],row['Long']],popup=row['Launch Site']).add_to(site_map)
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>

Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


Task 2: Mark the success/failed launches for each site on the map
No they are not.
Yes a lot of sites are close to the coast.

Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [10]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [11]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [12]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
site_map = folium.Map(location = [28.5,-80.6], zoom_start=5)

#Iterate through each launch record in dataframe
for index, row in spacex_df.iterrows():
    if row['class'] == 1:
        marker_color = 'green'
    else:
        marker_color = 'red'
    folium.Marker(location=[row['Lat'], row['Long']], popup=f"Site: {row['Launch Site']}<br>Outcome: {'Success' if row['class'] == 1 else 'Failure'}",
    icon=folium.Icon(color=marker_color)).add_to(site_map)

#display map
site_map

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [13]:
## Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed,
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # Access the launch site coordinates and class from the current record
    launch_site_lat = record['Lat']
    launch_site_lon = record['Long']
    launch_outcome = record['class']

    marker_color = 'green' if launch_outcome == 1 else 'red'

    # Create and add a Marker cluster to the site map
    marker = folium.Marker(
        location=[launch_site_lat, launch_site_lon],
        popup=f"Site: {record['Launch Site']}<br>Outcome: {'Success' if launch_outcome == 1 else 'Failure'}",
        icon=folium.Icon(color=marker_color)
    )
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [ ]:
# TASK 3: Calculate the distances between a launch site to its proximities

Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [14]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [15]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [ ]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [ ]:

# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [16]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
# Example: Find the distance from a launch site to a city

# Choose a launch site (e.g., KSC LC-39A)
launch_site_lat = 28.573255
launch_site_lon = -80.646895

# Coordinate of a city point (replace with your observed coordinates from MousePosition)
city_lat = 28.6114
city_lon = -80.8208 # Example coordinates for Titusville, FL

# Calculate the distance
distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)

print(f"Distance from KSC LC-39A to city: {distance_city:.2f} KM")

# Create a marker for the city point with the distance
distance_marker_city = folium.Marker(
   [city_lat, city_lon],
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_city:.2f} KM</b></div>',
       )
   )
site_map.add_child(distance_marker_city)

# Create a `folium.PolyLine` object between the launch site and city
city_coordinates = [[launch_site_lat, launch_site_lon], [city_lat, city_lon]]
lines_city=folium.PolyLine(locations=city_coordinates, weight=1)
site_map.add_child(lines_city)

# Display the map
site_map

Distance from KSC LC-39A to city: 17.51 KM


In [17]:
# Example: Find the distance from a launch site to a railway

# Choose a launch site (e.g., KSC LC-39A)
launch_site_lat = 28.573255
launch_site_lon = -80.646895

# Coordinate of a railway point (replace with your observed coordinates from MousePosition)
railway_lat = 28.57201
railway_lon = -80.58528 # Example coordinates for a nearby railway

# Calculate the distance
distance_railway = calculate_distance(launch_site_lat, launch_site_lon, railway_lat, railway_lon)

print(f"Distance from KSC LC-39A to railway: {distance_railway:.2f} KM")

# Create a marker for the railway point with the distance
distance_marker_railway = folium.Marker(
   [railway_lat, railway_lon],
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_railway:.2f} KM</b></div>',
       )
   )
site_map.add_child(distance_marker_railway)

# Create a `folium.PolyLine` object between the launch site and railway
railway_coordinates = [[launch_site_lat, launch_site_lon], [railway_lat, railway_lon]]
lines_railway=folium.PolyLine(locations=railway_coordinates, weight=1)
site_map.add_child(lines_railway)

# Display the map (optional, you can display after adding all elements)
# site_map

Distance from KSC LC-39A to railway: 6.02 KM


In [18]:
# Example: Find the distance from a launch site to a highway

# Choose a launch site (e.g., KSC LC-39A)
launch_site_lat = 28.573255
launch_site_lon = -80.646895

# Coordinate of a highway point (replace with your observed coordinates from MousePosition)
highway_lat = 28.5648
highway_lon = -80.5778 # Example coordinates for a nearby highway

# Calculate the distance
distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)

print(f"Distance from KSC LC-39A to highway: {distance_highway:.2f} KM")

# Create a marker for the highway point with the distance
distance_marker_highway = folium.Marker(
   [highway_lat, highway_lon],
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_highway:.2f} KM</b></div>',
       )
   )
site_map.add_child(distance_marker_highway)

# Create a `folium.PolyLine` object between the launch site and highway
highway_coordinates = [[launch_site_lat, launch_site_lon], [highway_lat, highway_lon]]
lines_highway=folium.PolyLine(locations=highway_coordinates, weight=1)
site_map.add_child(lines_highway)

# Display the map
site_map

Distance from KSC LC-39A to highway: 6.81 KM


After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.




*   Launch site is about 6.02 kms away from the railways.
*   Closest highway is about 6.8 kms away to launch site.
*   Coast line is the closest as it's just about 0.93 kms away!
*   They surely seem to! The site is 17.5kms away from the city.




# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
